File: android/gemcore/src/main/kotlin/com/gemwallet/android/model/PaymentTransfer.kt
package com.gemwallet.android.model

import com.gemwallet.android.domains.asset.chain
import com.gemwallet.android.ext.checksumAddress
import com.gemwallet.android.ext.isMemoSupport
import com.gemwallet.android.ext.isValidAddress
import com.wallet.core.primitives.PaymentAmount
import com.wallet.core.primitives.PaymentRequest
import java.math.BigInteger

class PaymentTransfer(private val assetInfo: AssetInfo) {

    fun destination(request: PaymentRequest): PaymentDestination.Transfer {
        val assetId = request.assetId
        if (assetId != null && assetId != assetInfo.asset.id) {
            return PaymentDestination.Unsupported
        }
        val params = if (requiresMemo(request)) null else decodedTransfer(request)

        return when (params) {
            null -> PaymentDestination.Recipient(assetInfo.asset.id, request)
            else -> PaymentDestination.Confirm(params)
        }
    }

    fun decodedTransfer(request: PaymentRequest): ConfirmParams.TransferParams? {
        val asset = assetInfo.asset
        val owner = assetInfo.owner ?: return null
        val assetId = request.assetId
        if (assetId != null && assetId != asset.id) {
            return null
        }
        val value = transferValue(request) ?: return null
        val address = asset.chain.checksumAddress(request.address)

        if (!asset.chain.isValidAddress(address)) {
            return null
        }
        return ConfirmParams.Builder(asset, owner, value)
            .transfer(DestinationAddress(address), request.memo, request.references.orEmpty())
    }

    private fun requiresMemo(request: PaymentRequest): Boolean =
        assetInfo.asset.chain.isMemoSupport() && request.memo.isNullOrEmpty()

    private fun transferValue(request: PaymentRequest): BigInteger? =
        when (val amount = request.amount) {
            is PaymentAmount.ExactValue -> Crypto.exact(amount.content, assetInfo.asset.decimals)?.atomicValue
            is PaymentAmount.AtomicValue -> amount.content.toBigIntegerOrNull()
            null -> null
        }
}

File: ios/Packages/PrimitivesComponents/Sources/Types/PaymentTransfer.swift
// Copyright (c). Gem Wallet. All rights reserved.

import BigInt
import Formatters
import GemstonePrimitives
import Localization
import Primitives

public struct PaymentTransfer: Sendable {
    public enum Destination: Sendable {
        case confirm(TransferData)
        case recipient(RecipientData)
    }

    private let asset: Asset

    public init(asset: Asset) {
        self.asset = asset
    }

    public func destination(for payment: PaymentRequest) throws -> Destination {
        if let data = try confirmation(for: payment, type: .transfer(asset)) {
            return .confirm(data)
        }
        return .recipient(
            RecipientData(
                recipient: try recipient(for: payment),
                amount: payment.exactAmount,
            ),
        )
    }
}

extension PaymentTransfer {
    func confirmation(for payment: PaymentRequest, type: TransferDataType) throws -> TransferData? {
        let recipient = try recipient(for: payment)
        guard let value = confirmableValue(of: payment, address: recipient.address) else {
            return nil
        }
        return TransferData(
            type: type,
            recipientData: RecipientData(recipient: recipient, amount: nil),
            amount: .exact(value),
        )
    }
}

// MARK: - Private

private extension PaymentTransfer {
    func recipient(for payment: PaymentRequest) throws -> Recipient {
        guard payment.assetId == nil || payment.assetId == asset.id else {
            throw AnyError(Localized.Errors.notSupported)
        }
        return Recipient(
            name: nil,
            address: asset.chain.checksumAddress(payment.address),
            memo: payment.memo,
        )
    }

    func confirmableValue(of payment: PaymentRequest, address: String) -> BigInt? {
        guard asset.chain.isValidAddress(address) else { return .none }
        if asset.chain.isMemoSupported {
            guard payment.memo?.isEmpty == false else { return .none }
        }
        return transferValue(of: payment)
    }

    func transferValue(of payment: PaymentRequest) -> BigInt? {
        switch payment.amount {
        case let .exactValue(value): try? BigNumberFormatter.standard.exactNumber(from: value, decimals: asset.decimals.asInt)
        case let .atomicValue(value): BigInt(value)
        case .none: .none
        }
    }
}
